proposals/sealed-inline-classes.md

```diff
@@ -0,0 +1,593 @@
+
+- Sealed inline classes cannot have primary constructors.
+
+- If sealed inline class is a child of another sealed inline class, there can be no other inline children, including other sealed inline class, since sealed inline classes are mapped to `Any?`, which is open class, which other classes override and there is no way to distinguish `Any?` from other type.
+
+- Value objects must be a child of sealed inline class.
+
+- Value objects cannot be annotated with `@JvmInline`.
+
+- Other restrictions of inline classes also apply to sealed inline classes.
+  - Sealed inline classes can only implement interfaces.
+  - Sealed inline classes and inline children cannot have backing fields.
+  - Sealed inline classes and their children do not support referencial equality (`===`).
+  - Sealed inline classes and inline children cannot be local or inner.
+
+## JVM representation
+All sealed inline classes are mapped to `Any?`. The compiler generates synthetic field for the underlying value:
+
+```kotlin
+@JvmInline
+sealed value class IC
+```
+
+becomes:
+
+```kotlin
+class IC {
+  synthetic private val $value: Any?
+  
+  synthetic public fun $value(): Any? = $value
+}
+```
+
+Note, that we generate the getter for the field - it is called in static `-impl`
+function replacements of inline children's methods. 
+
+There can be no boxed inline children - thus, original methods are removed. In
+usual inline classes they redirect to static replacement functions.
+
+If one of children has primitive underlying value, the value is boxed:
+```kotlin
+@JvmInline
+sealed value class IC
+
+@JvmInline
+value class ICInt(val i: Int): IC()
+
+value object ICObject: IC()
+
+val ic: IC = ICInt(1)
+```
+
+the compiler generates the following code
+
+```kotlin
+val ic = Integer.valueOf(1)
+```
+
+The following rules for passing to function and returning from the function apply:
+
+- When we pass sealed inline class, it is mapped to `Any?`:
+```kotlin
+fun foo(ic: IC)
+```
+
+becomes:
+```kotlin
+fun foo-<hash>(ic: Any?)
+```
+
+`<hash>` is computed using usual [inline classes mangling rules](https://github.com/Kotlin/KEEP/blob/master/proposals/inline-classes.md#mangling-rules).
+
+- When we pass nullable sealed inline class, it is mapped to boxed sealed inline class:
+```kotlin
+fun foo(ic: IC?)
+```
+
+becomes:
+```kotlin
+fun foo-<hash>(ic: IC?)
+```
+
+- When we pass inline child, it is mapped to reference underlying type:
+```kotin
+fun foo(icInt: ICInt)
+```
+
+becomes:
+```kotlin
+fun foo-<hash>(icInt: Int?)
+```
+
+- When we pass nullable inline child, it is mapped to sealed inline class:
+```kotlin
+fun foo(icInt: ICInt?)
+```
+
+becomes:
+```kotlin
+fun foo-<hash>(icInt: IC?)
+```
+
+Since there can be no boxed inline children.
+
+- When we pass noinline child, nothing special happens:
+```kotlin
+fun foo(icObject: ICObject)
+``` 
+
+becomes:
+```kotlin
+fun foo(icObject: ICObject)
+```
+
+## Is checks
+Consider the following example
```

Comment:
This example does not compile because `sealed inline class child cannot be distinguished from another one`